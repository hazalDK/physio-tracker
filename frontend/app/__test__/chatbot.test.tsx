import React from "react";
import { render, fireEvent, waitFor, act } from "@testing-library/react-native";
import renderer from "react-test-renderer";
import Chatbot from "../chatbot";

// Mock dependencies
jest.mock("axios");
jest.mock("@/hooks/useAuth", () => {
  return {
    useAuth: jest.fn().mockImplementation(() => ({
      createApiInstance: jest.fn(),
      refreshToken: jest.fn(),
    })),
  };
});

jest.mock("@react-navigation/native", () => ({
  ...jest.requireActual("@react-navigation/native"),
  useNavigation: () => ({
    goBack: jest.fn(),
  }),
}));

jest.mock("react-native-root-toast", () => ({
  show: jest.fn(),
  durations: { SHORT: 2000 },
  positions: { BOTTOM: 50 },
}));

jest.mock("@expo/vector-icons", () => ({
  Ionicons: "Ionicons",
}));

// Don't mock the useRef hook - let it use the real implementation
jest.mock("react", () => {
  const actualReact = jest.requireActual("react");
  return {
    ...actualReact,
    // We'll keep the real useRef implementation
  };
});

// Access to the mocked useAuth function
const mockUseAuth = jest.requireMock("@/hooks/useAuth").useAuth;

describe("Chatbot Component", () => {
  beforeEach(() => {
    jest.clearAllMocks();

    // Setup API mocks
    const mockApi = {
      get: jest.fn(),
      post: jest.fn(),
    };

    mockApi.get.mockImplementation((url) => {
      if (url === "/user-exercises/") {
        return Promise.resolve({ data: [] });
      }
      if (url === "/reports/") {
        return Promise.resolve({ data: [] });
      }
      return Promise.reject(new Error("Unknown URL"));
    });

    mockApi.post.mockImplementation((url) => {
      if (url === "/api/chatbot/") {
        return Promise.resolve({
          data: { message: "Bot response" },
        });
      }
      if (url === "/api/reset-chat/") {
        return Promise.resolve({});
      }
      return Promise.reject(new Error("Unknown URL"));
    });

    // Update useAuth mock implementation
    mockUseAuth.mockImplementation(() => ({
      createApiInstance: jest.fn().mockResolvedValue(mockApi),
      refreshToken: jest.fn().mockResolvedValue("new-token"),
    }));
  });

  it("renders correctly with data", async () => {
    let tree: any;
    await act(async () => {
      tree = renderer.create(<Chatbot />);
    });

    // Allow any pending state updates and async operations to complete
    await new Promise((resolve) => setTimeout(resolve, 0));

    // Then call toJSON() outside of act
    const treeJSON = tree.toJSON();
    expect(treeJSON).not.toBeNull();
    expect(treeJSON).toMatchSnapshot();
  });

  it("renders initial messages and loading state", async () => {
    const { getByText, queryByTestId, findByText } = render(<Chatbot />);

    // Initially should show loading indicator
    expect(queryByTestId("activity-indicator")).toBeTruthy();

    // Wait for welcome message to appear - using findByText which has built-in waiting
    const welcomeMessage = await findByText(
      "Welcome to your physiotherapy assistant! How can I help you today?"
    );
    expect(welcomeMessage).toBeTruthy();

    // Verify DISCLAIMER is also shown
    expect(
      getByText(
        "DISCLAIMER: This feedback is generated by AI and is not a substitute for professional medical advice. Always consult with your healthcare provider before starting, changing, or stopping any exercise program."
      )
    ).toBeTruthy();

    // Now loading indicator should be gone
    expect(queryByTestId("activity-indicator")).toBeFalsy();
  });

  it("sends user message and displays bot response", async () => {
    const { getByPlaceholderText, getByTestId, getByText, findByText } = render(
      <Chatbot />
    );

    // Wait for the component to finish loading and show welcome message
    await findByText(
      "Welcome to your physiotherapy assistant! How can I help you today?"
    );

    // Type a message
    const input = getByPlaceholderText("Type your message...");
    fireEvent.changeText(input, "Hello bot");

    // Send the message
    const sendButton = getByTestId("send-button");
    fireEvent.press(sendButton);

    // Should show user message
    expect(getByText("Hello bot")).toBeTruthy();

    // Wait for bot response using findByText instead of manual waitFor
    const botResponse = await findByText("Bot response");
    expect(botResponse).toBeTruthy();
  });

  it("handles API errors when sending messages", async () => {
    // Mock API error
    const mockApi = {
      get: jest.fn().mockResolvedValue({ data: [] }),
      post: jest.fn().mockRejectedValue(new Error("API error")),
    };

    mockUseAuth.mockImplementation(() => ({
      createApiInstance: jest.fn().mockResolvedValue(mockApi),
      refreshToken: jest.fn().mockResolvedValue("new-token"),
    }));

    const { getByPlaceholderText, getByTestId, findByText } = render(
      <Chatbot />
    );

    // Wait for welcome message
    await findByText(
      "Welcome to your physiotherapy assistant! How can I help you today?"
    );

    // Type and send message
    const input = getByPlaceholderText("Type your message...");
    fireEvent.changeText(input, "This will fail");
    const sendButton = getByTestId("send-button");
    fireEvent.press(sendButton);

    // Wait for error message
    const errorMessage = await findByText(
      "Sorry, I encountered an error. Please try again."
    );
    expect(errorMessage).toBeTruthy();
  });

  it("resets chat history", async () => {
    // Set up the API mock to respond appropriately to reset
    const mockResetApi = jest.fn().mockResolvedValue({});

    const mockApi = {
      get: jest.fn().mockResolvedValue({ data: [] }),
      post: jest.fn().mockImplementation((url) => {
        if (url === "/api/chatbot/") {
          return Promise.resolve({ data: { message: "Bot response" } });
        }
        if (url === "/api/reset-chat/") {
          return mockResetApi();
        }
        return Promise.reject(new Error("Unknown URL"));
      }),
    };

    // Ensure we're using fresh mocks
    mockUseAuth.mockImplementation(() => ({
      createApiInstance: jest.fn().mockResolvedValue(mockApi),
      refreshToken: jest.fn().mockResolvedValue("new-token"),
    }));

    const { getByPlaceholderText, findByText, queryByText, getByTestId } =
      render(<Chatbot />);

    // Wait for welcome message
    await findByText(
      "Welcome to your physiotherapy assistant! How can I help you today?"
    );

    // Send a message and wait for response
    const input = getByPlaceholderText("Type your message...");
    fireEvent.changeText(input, "Test message");
    fireEvent.press(getByTestId("send-button"));
    await findByText("Bot response");

    // Now press reset
    fireEvent.press(getByTestId("reset-button"));

    // Verify the reset API was called
    await waitFor(() => expect(mockResetApi).toHaveBeenCalled());

    // Wait for reset to complete and component to re-render
    await findByText(
      "Welcome to your physiotherapy assistant! How can I help you today?"
    );

    // Test messages should be gone
    expect(queryByText("Test message")).toBeFalsy();
  });

  it("reset chat history fails", async () => {
    // Mock API error for reset but success for message
    const mockApi = {
      get: jest.fn().mockResolvedValue({ data: [] }),
      post: jest.fn().mockImplementation((url) => {
        if (url === "/api/chatbot/") {
          return Promise.resolve({
            data: { message: "Bot response" },
          });
        }
        if (url === "/api/reset-chat/") {
          return Promise.reject(new Error("Reset API error"));
        }
        return Promise.reject(new Error("Unknown URL"));
      }),
    };

    mockUseAuth.mockImplementation(() => ({
      createApiInstance: jest.fn().mockResolvedValue(mockApi),
      refreshToken: jest.fn().mockResolvedValue("new-token"),
    }));

    const { getByPlaceholderText, getByTestId, findByText } = render(
      <Chatbot />
    );

    // Wait for welcome message
    await findByText(
      "Welcome to your physiotherapy assistant! How can I help you today?"
    );

    // Type and send message
    const input = getByPlaceholderText("Type your message...");
    fireEvent.changeText(input, "Test message");
    const sendButton = getByTestId("send-button");
    fireEvent.press(sendButton);

    // Wait for bot response
    await findByText("Bot response");

    // Mock console.error to avoid test noise
    const originalConsoleError = console.error;
    console.error = jest.fn();

    try {
      // Press reset button
      const resetButton = getByTestId("reset-button");
      fireEvent.press(resetButton);

      // Wait for Alert to be called (we need to mock Alert separately)
      await waitFor(() => {
        expect(mockApi.post).toHaveBeenCalledWith("/api/reset-chat/", {});
      });
    } finally {
      // Restore console.error
      console.error = originalConsoleError;
    }
  });
});
