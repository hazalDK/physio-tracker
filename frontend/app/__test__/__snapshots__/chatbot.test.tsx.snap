// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chatbot Component renders correctly with data 1`] = `
<RCTSafeAreaView
  style={
    {
      "backgroundColor": "rgba(255, 255, 255, 1)",
      "flexBasis": "0%",
      "flexGrow": 1,
      "flexShrink": 1,
    }
  }
>
  <View
    onLayout={[Function]}
    style={
      [
        {
          "flexBasis": "0%",
          "flexGrow": 1,
          "flexShrink": 1,
        },
        {
          "paddingBottom": 0,
        },
      ]
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "borderBottomColor": "rgba(229, 231, 235, 1)",
          "borderBottomWidth": 1,
          "borderLeftColor": "rgba(229, 231, 235, 1)",
          "borderRightColor": "rgba(229, 231, 235, 1)",
          "borderTopColor": "rgba(229, 231, 235, 1)",
          "flexDirection": "row",
          "justifyContent": "space-between",
          "paddingBottom": 12,
          "paddingLeft": 16,
          "paddingRight": 16,
          "paddingTop": 12,
        }
      }
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          [
            {
              "borderBottomLeftRadius": 9999,
              "borderBottomRightRadius": 9999,
              "borderTopLeftRadius": 9999,
              "borderTopRightRadius": 9999,
              "paddingBottom": 8,
              "paddingLeft": 8,
              "paddingRight": 8,
              "paddingTop": 8,
            },
            {
              "opacity": 1,
            },
          ]
        }
      >
        <Ionicons
          color="#14b8a6"
          name="arrow-back"
          size={24}
        />
      </View>
      <Text
        style={
          {
            "color": "rgba(31, 41, 55, 1)",
            "fontSize": 18,
            "fontWeight": "600",
            "lineHeight": 28,
          }
        }
      >
        Physio Assistant
      </Text>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          [
            {
              "borderBottomLeftRadius": 9999,
              "borderBottomRightRadius": 9999,
              "borderTopLeftRadius": 9999,
              "borderTopRightRadius": 9999,
              "paddingBottom": 8,
              "paddingLeft": 8,
              "paddingRight": 8,
              "paddingTop": 8,
            },
            {
              "opacity": 1,
            },
          ]
        }
        testID="reset-button"
      >
        <Ionicons
          color="#14b8a6"
          name="refresh"
          size={24}
        />
      </View>
    </View>
    <RCTScrollView
      contentContainerStyle={
        {
          "paddingBottom": 16,
        }
      }
      onContentSizeChange={[Function]}
      style={
        {
          "flexBasis": "0%",
          "flexGrow": 1,
          "flexShrink": 1,
          "paddingLeft": 16,
          "paddingRight": 16,
          "paddingTop": 16,
        }
      }
    >
      <View>
        <View
          style={
            {
              "alignSelf": "flex-start",
              "backgroundColor": "rgba(254, 243, 199, 1)",
              "borderBottomColor": "rgba(252, 211, 77, 1)",
              "borderBottomLeftRadius": 8,
              "borderBottomRightRadius": 8,
              "borderBottomWidth": 1,
              "borderLeftColor": "rgba(252, 211, 77, 1)",
              "borderLeftWidth": 1,
              "borderRightColor": "rgba(252, 211, 77, 1)",
              "borderRightWidth": 1,
              "borderTopColor": "rgba(252, 211, 77, 1)",
              "borderTopLeftRadius": 8,
              "borderTopRightRadius": 8,
              "borderTopWidth": 1,
              "marginBottom": 16,
              "paddingBottom": 8,
              "paddingLeft": 12,
              "paddingRight": 12,
              "paddingTop": 8,
            }
          }
        >
          <Text
            style={
              {
                "color": "rgba(146, 64, 14, 1)",
                "fontSize": 12,
                "fontWeight": "500",
                "lineHeight": 16,
              }
            }
          >
            DISCLAIMER: This feedback is generated by AI and is not a substitute for professional medical advice. Always consult with your healthcare provider before starting, changing, or stopping any exercise program.
          </Text>
        </View>
        <View
          style={
            [
              {
                "alignSelf": "flex-start",
                "backgroundColor": "rgba(229, 231, 235, 1)",
                "borderBottomLeftRadius": 0,
                "borderBottomRightRadius": 16,
                "borderTopLeftRadius": 16,
                "borderTopRightRadius": 16,
                "marginBottom": 8,
                "paddingBottom": 12,
                "paddingLeft": 16,
                "paddingRight": 16,
                "paddingTop": 12,
              },
              {
                "maxWidth": "75%",
              },
            ]
          }
        >
          <Text
            style={
              {
                "color": "rgba(31, 41, 55, 1)",
              }
            }
          >
            Welcome to your physiotherapy assistant! How can I help you today?
          </Text>
        </View>
      </View>
    </RCTScrollView>
    <View
      style={
        {
          "alignItems": "center",
          "borderBottomColor": "rgba(229, 231, 235, 1)",
          "borderLeftColor": "rgba(229, 231, 235, 1)",
          "borderRightColor": "rgba(229, 231, 235, 1)",
          "borderTopColor": "rgba(229, 231, 235, 1)",
          "borderTopWidth": 1,
          "flexDirection": "row",
          "paddingBottom": 8,
          "paddingLeft": 16,
          "paddingRight": 16,
          "paddingTop": 8,
        }
      }
    >
      <TextInput
        multiline={true}
        onChangeText={[Function]}
        placeholder="Type your message..."
        style={
          {
            "backgroundColor": "rgba(243, 244, 246, 1)",
            "borderBottomLeftRadius": 9999,
            "borderBottomRightRadius": 9999,
            "borderTopLeftRadius": 9999,
            "borderTopRightRadius": 9999,
            "color": "rgba(0, 0, 0, 1)",
            "flexBasis": "0%",
            "flexGrow": 1,
            "flexShrink": 1,
            "marginRight": 8,
            "paddingBottom": 8,
            "paddingLeft": 16,
            "paddingRight": 16,
            "paddingTop": 8,
          }
        }
        value=""
      />
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": true,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          [
            {
              "borderBottomLeftRadius": 9999,
              "borderBottomRightRadius": 9999,
              "borderTopLeftRadius": 9999,
              "borderTopRightRadius": 9999,
              "paddingBottom": 8,
              "paddingLeft": 8,
              "paddingRight": 8,
              "paddingTop": 8,
            },
            {
              "backgroundColor": "#94a3b8",
              "opacity": 1,
            },
          ]
        }
        testID="send-button"
      >
        <Ionicons
          color="white"
          name="send"
          size={22}
        />
      </View>
    </View>
  </View>
</RCTSafeAreaView>
`;
